cache/lru/lru.go

```diff
@@ -0,0 +1,163 @@
+package lru
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+
+	"github.com/lightninglabs/neutrino/cache"
+)
+
+// elementMap is an alias for a map from a generic interface to a list.Element.
+type elementMap map[interface{}]*list.Element
+
+// entry represents a (key,value) pair entry in the Cache. The Cache's list
+// stores entries which let us get the cache key when an entry is evicted.
+type entry struct {
+	key   interface{}
+	value cache.Value
+}
+
+// Cache provides a generic thread-safe lru cache that can be used for
+// storing filters, blocks, etc.
+type Cache struct {
+	// capacity represents how much this cache can hold. It could be number of
+	// elements or a number of bytes, decided by the cache.Value's Size.
+	capacity uint64
+
+	// size represents the size of all the elements currenty in the cache.
+	size uint64
+
+	// ll is a doubly linked list which keeps track of recency of used elements
+	// by moving them to the front.
+	ll *list.List
+
+	// cache is a generic cache which allows us to find an elements position
+	// in the ll list from a given key.
+	cache elementMap
+
+	// mtx is used to make sure the Cache is thread-safe.
+	mtx sync.RWMutex
+}
+
+// NewCache return a cache with specified capacity, the cache's size can't
+// exceed that given capacity.
+func NewCache(capacity uint64) *Cache {
+	return &Cache{
+		capacity: capacity,
+		ll:       list.New(),
+		cache:    make(elementMap),
+	}
+}
+
+// evict will evict as many elements as necessary to make enough space for a new
+// element with size needed to be inserted.
+func (c *Cache) evict(needed uint64) error {
+	if needed > c.capacity {
+		return fmt.Errorf(
+			"can't evict %v elements in size, since capacity is %v",
+			needed, c.capacity)
+	}
+
+	room := c.capacity - c.size
+	for room < needed {
+		// We still need to evict some more elements.
+		if c.ll.Len() == 0 {
+			// We should never reach here.
+			return fmt.Errorf(
+				"all elements got evicted, yet still need to evict %v,"+
+					"likelihood of error during size calculation",
+				needed-room)
+		}
+
+		// Find the least recently used item.
+		if elr := c.ll.Back(); elr != nil {
+			// Determine lru item's size.
+			ce := elr.Value.(*entry)
+			es, err := ce.value.Size()
+			if err != nil {
+				return fmt.Errorf(
+					"couldn't determine size of existing cache value %v", err)
+			}
+
+			// Account for that element's removal in evicted and cache size.
+			c.size -= es
+			room = c.capacity - c.size
```

Comment:
nit: this can be simplified to making the loop
```go
// We'll evict elements until the cache has room for an element 
// of size "needed".
for c.capacity - c.size < needed {
```